plugin/manager.go

```diff
@@ -0,0 +1,318 @@
+			err = inst.Disable()
+			m.mutex.Unlock()
+			if err != nil {
+				return err
+			}
+		}
+		delete(m.instances, pluginConf.ID)
+	}
+	return nil
+}
+
+func (m *Manager) loadPlugins(directory string) error {
+	if directory == "" {
+		return nil
+	}
+
+	pluginFiles, err := ioutil.ReadDir(directory)
+	if err != nil {
+		return fmt.Errorf("error while reading directory %s", err)
+	}
+	for _, f := range pluginFiles {
+		pluginPath := filepath.Join(directory, "./", f.Name())
+		pRaw, err := plugin.Open(pluginPath)
+		if err != nil {
+			return fmt.Errorf("error while opening plugin %s; %s", pluginPath, err)
+		}
+		compatPlugin, err := compat.Wrap(pRaw)
+		if err != nil {
+			return err
+		}
+		if err := m.LoadPlugin(compatPlugin); err != nil {
+			return nil
+		}
+	}
+	return nil
+}
+
+// LoadPlugin loads a compat plugin, exported to sideload plugins for testing purposes
+func (m *Manager) LoadPlugin(compatPlugin compat.Plugin) error {
+	modulePath := compatPlugin.PluginInfo().ModulePath
+	if _, ok := m.plugins[modulePath]; ok {
+		return fmt.Errorf("plugin with module path %s is present at least twice", modulePath)
+	}
+	m.plugins[modulePath] = compatPlugin
+	return nil
+}
+
+// InitializeForUserID initializes all plugin instances for a given user
+func (m *Manager) InitializeForUserID(userID uint) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	user := m.db.GetUserByID(userID)
+	if user != nil {
+		return m.initializeForUser(*user)
+	}
+	return fmt.Errorf("user with id %d not found", userID)
+}
+
+func (m *Manager) initializeForUser(user model.User) error {
+
+	userCtx := compat.UserContext{
+		ID:    user.ID,
+		Name:  user.Name,
+		Admin: user.Admin,
+	}
+
+	for _, p := range m.plugins {
+		if err := m.initializeSingleUserPlugin(userCtx, p); err != nil {
+			return err
+		}
+	}
+
+	for _, app := range m.db.GetApplicationsByUser(user.ID) {
+		if conf := m.db.GetPluginConfByApplicationID(app.ID); conf != nil {
+			_, compatExist := m.plugins[conf.ModulePath]
+			app.Internal = compatExist
+		} else {
+			app.Internal = false
+		}
+		m.db.UpdateApplication(app)
+	}
+
+	return nil
+}
+
+func (m *Manager) initializeSingleUserPlugin(userCtx compat.UserContext, p compat.Plugin) error {
+	info := p.PluginInfo()
+	instance := p.NewPluginInstance(userCtx)
+	userID := userCtx.ID
+
+	pluginConf := m.db.GetPluginConfByUserAndPath(userID, info.ModulePath)
+
+	if pluginConf == nil {
+		var err error
+		pluginConf, err = m.createPluginConf(instance, info, userID)
+		if err != nil {
+			return err
+		}
+	}
+
+	m.instances[pluginConf.ID] = instance
+
+	if compat.HasSupport(instance, compat.Messenger) {
+		instance.SetMessageHandler(redirectToChannel{
+			ApplicationID: pluginConf.ApplicationID,
+			UserID:        pluginConf.UserID,
+			Messages:      m.messages})
+	}
+	if compat.HasSupport(instance, compat.Storager) {
+		instance.SetStorageHandler(dbStorageHandler{pluginConf.ID, m.db})
+	}
+	if compat.HasSupport(instance, compat.Configurer) {
```

Comment:
Should we in some way provide a way for plugin to update its config scheme seamlessly? I think here we have several options:

- Leave it as is, the plugin updates configuration scheme by adding more fields and deprecating existing fields in its `Config` struct. The user will have to update configuration scheme manually. If `ValidateAndSetConfig` returns an error during initialization, the configuration is automatically replaced by the new `DefaultConfig`
```golang
const pluginVersion = 1
type Conf struct {
    SomeKey string
}
```
```golang
const pluginVersion = 2
type Conf struct {
    SomeKey        string
    SomeFeature bool
}
```
- Record a config scheme number along with the config itself for the plugin, and the plugin implement an interface like `ConfigurerUpgrader`. The interface implementation would look like this:
```golang
func (c *Plugin) ConfigSchema() uint {
    return 8
}

func (c *Plugin) UpgradeConfig(schema uint, yamlConf []byte) interface{} {
    if canUpgradeFrom(schema) {
        return upgradeConfig(schema, yamlConf)
    }
    return c.DefaultConfig()
}
```

Reply:
Thats not an easy question, I think that it will be overkill if we add config migration.

How about we do it like this:
* admin updates plugin which has a different config object.
* on startup gotify/server will disable all plugins with an invalid config
* users then need to fix the config before they are able to enable the plugin again

Reply:
I think that's a good idea. My expected behavior is:
1. Hit an error during Unmarshaling or ValidateAndSetConfig in initialization
2. Obtain a new DefaultConfig
3. Comment out existing configs and append to DefaultConfig
4. Use that as the new config
5. Disable the plugin